packages/blocks/src/lib/navigation-map/components/draw-tool.svelte

```diff
@@ -0,0 +1,105 @@
+<!-- 
+  @component
+  Renders an in-progress rectangle that represents a drawn area by the user.
+-->
+<script lang="ts">
+import * as THREE from 'three';
+import { T, createRawEventDispatcher } from '@threlte/core';
+import { useMapLibreEvent, useMapLibre } from '$lib';
+import { MercatorCoordinate, LngLat, type MapMouseEvent } from 'maplibre-gl';
+import { view } from '../stores';
+
+interface $$Events extends Record<string, unknown> {
+  /** Fires when a rectangle is drawn. */
+  update: { width: number; height: number; center: LngLat };
+}
+
+const dispatch = createRawEventDispatcher<$$Events>();
+const { map } = useMapLibre();
+
+let downLngLat = new LngLat(0, 0);
+let down = new MercatorCoordinate(0, 0, 0);
+
+let drawing = false;
+let width = 0;
+let height = 0;
+
+const moveSign = { x: 0, y: 0 };
+
+const toPrecisionLevel = (number: number, decimals: number) => {
+  const multiplier = 10 ** decimals;
+  return Math.floor(number * multiplier) / multiplier;
+};
+
+const handlePointerMove = (event: MapMouseEvent) => {
+  const move = MercatorCoordinate.fromLngLat(event.lngLat, 0);
+  const scale = move.meterInMercatorCoordinateUnits();
+
+  moveSign.x = Math.sign(move.x - down.x);
+  moveSign.y = Math.sign(move.y - down.y);
+
+  width = toPrecisionLevel(Math.abs(move.x - down.x) / scale, 2);
+  height = toPrecisionLevel(Math.abs(move.y - down.y) / scale, 2);
+};
+
+const handlePointerUp = () => {
+  drawing = false;
+
+  const scale = down.meterInMercatorCoordinateUnits();
+  const offset = new MercatorCoordinate(
+    -moveSign.x * (width / 2) * scale,
+    -moveSign.y * (height / 2) * scale
+  );
+
+  down.x -= offset.x;
+  down.y -= offset.y;
+
+  const center = down.toLngLat();
+
+  dispatch('update', { width, height, center });
+
+  width = 0;
+  height = 0;
+};
+
+const handleGeometryCreate = ({ ref }: { ref: THREE.BufferGeometry }) => {
+  ref.rotateX(-Math.PI / 2);
+};
+
+useMapLibreEvent('mousedown', (event) => {
+  if (event.originalEvent.shiftKey) {
+    event.preventDefault();
+    drawing = true;
+    downLngLat = event.lngLat;
+    down = MercatorCoordinate.fromLngLat(event.lngLat, 0);
+  }
+});
+
+$: if (drawing) {
+  map.on('mousemove', handlePointerMove);
+  map.on('mouseup', handlePointerUp);
+} else {
+  map.off('mousemove', handlePointerMove);
+  map.off('mouseup', handlePointerUp);
+}
+</script>
+
+<T.Group userData.lngLat={downLngLat}>
+  <T.Mesh
+    position.x={(-moveSign.x * width) / 2}
+    position.z={(-moveSign.y * height) / 2}
+  >
+    {#if $view === '3D'}
+      <T.BoxGeometry
+        args={[width, height, 10]}
+        on:create={handleGeometryCreate}
+      />
+    {:else}
+      <T.PlaneGeometry
+        args={[width, height]}
+        on:create={handleGeometryCreate}
+      />
+    {/if}
+    <T is={new THREE.MeshPhongMaterial({ color: 'red' })} />
```

Comment:
Nit: red means error and generally sounds alarm bells for me when it is used in a UI. This doesn't feel like danger. Is there another color we can use?

Reply:
Agreed with this principle. The main reason why we chose red was to easily contrast with the map color palette. I'll revisit this though with Peter and Nat to see if we can come up with something better.